File: system_monitor/collector/runner.py
from __future__ import annotations

import threading
import time
from typing import Any, Callable

import psutil

from ..config_loader import SensorConfig, SensorsFile, load_sensors_config
from ..disk_discovery import discover_disk_sensors
from ..storage import MetricStore
from .base import Sensor, SensorReading
from .registry import create_sensor


class Collector:
    def __init__(
        self,
        store: MetricStore | None = None,
        on_update: Callable[[dict[str, Any]], None] | None = None,
        config: SensorsFile | None = None,
        config_loader: Callable[[], SensorsFile] | None = None,
    ) -> None:
        self.store = store
        self.on_update = on_update
        self._config_loader = config_loader or load_sensors_config
        self._stop = threading.Event()
        self._thread: threading.Thread | None = None
        self._config = config or self._config_loader()
        self._sensors: dict[str, Sensor] = {}
        self._last_poll: dict[str, float] = {}
        self._latest: dict[str, dict[str, Any]] = {}
        self._lock = threading.Lock()
        self.reload_config()

    def reload_config(self, config: SensorsFile | None = None) -> None:
        if config is not None:
            self._config = config
        else:
            self._config = self._config_loader()
        sensors: dict[str, Sensor] = {}
        configured_paths = {
            cfg.params.get("path")
            for cfg in self._config.sensors
            if cfg.type == "system.disk_usage" and cfg.params.get("path")
        }

        for item in self._config.sensors:
            if not item.enabled or not item.is_supported_on_platform():
                continue
            try:
                sensors[item.id] = create_sensor(item)
            except ValueError:
                continue

        if self._config.settings.auto_discover_disks:
            for item in discover_disk_sensors():
                mount = item.params.get("path")
                if mount in configured_paths:
                    continue
                try:
                    sensors[item.id] = create_sensor(item)
                except ValueError:
                    continue

        with self._lock:
            self._sensors = sensors

    def start(self) -> None:
        if self._thread and self._thread.is_alive():
            return
        psutil.cpu_percent(interval=None)
        self._stop.clear()
        self._thread = threading.Thread(target=self._run, name="system-monitor-collector", daemon=True)
        self._thread.start()

    def stop(self) -> None:
        self._stop.set()
        if self._thread:
            self._thread.join(timeout=2)

    def _interval_for(self, config: SensorConfig) -> int:
        return config.interval_sec or self._config.settings.default_interval_sec

    def _run(self) -> None:
        last_cleanup = 0.0
        while not self._stop.is_set():
            now = time.time()
            updates: dict[str, dict[str, Any]] = {}

            with self._lock:
                sensors = dict(self._sensors)

            for sensor_id, sensor in sensors.items():
                config = sensor.config
                interval = self._interval_for(config)
                last = self._last_poll.get(sensor_id, 0.0)
                if now - last < interval:
                    continue

                reading = self._safe_read(sensor)
                if self.store is not None:
                    self.store.insert(sensor_id, reading.value, reading.status, ts=now)
                payload = {
                    **reading.to_dict(),
                    "name": config.name,
                    "unit": config.unit,
                    "type": config.type,
                    "ts": now,
                }
                with self._lock:
                    self._latest[sensor_id] = payload
                updates[sensor_id] = payload
                self._last_poll[sensor_id] = now

            if updates and self.on_update:
                self.on_update(updates)

            if self.store is not None:
                retention = self._config.settings.retention_days
                if now - last_cleanup > 3600:
                    self.store.cleanup(retention)
                    last_cleanup = now

            time.sleep(0.5)

    def _safe_read(self, sensor: Sensor) -> SensorReading:
        try:
            return sensor.read()
        except Exception as exc:
            return SensorReading(
                sensor_id=sensor.id,
                value=None,
                status="error",
                error=str(exc),
            )

    def get_latest_snapshot(self) -> dict[str, dict[str, Any]]:
        with self._lock:
            return dict(self._latest)

    def get_sensor_configs(self) -> list[SensorConfig]:
        configs = list(self._config.sensors)
        configured_paths = {
            cfg.params.get("path")
            for cfg in configs
            if cfg.type == "system.disk_usage" and cfg.params.get("path")
        }
        for item in discover_disk_sensors():
            if item.params.get("path") in configured_paths:
                continue
            configs.append(item)
        return configs

    def get_settings(self) -> dict[str, Any]:
        return self._config.settings.model_dump(mode="json")

    def get_active_sensor_ids(self) -> set[str]:
        with self._lock:
            return set(self._sensors.keys())

    def build_sensor_meta(self) -> list[dict[str, Any]]:
        active_ids = self.get_active_sensor_ids()
        items = []
        for cfg in self.get_sensor_configs():
            if cfg.id not in active_ids:
                continue
            items.append(
                {
                    "id": cfg.id,
                    "name": cfg.name,
                    "type": cfg.type,
                    "unit": cfg.unit,
                    "enabled": cfg.enabled,
                    "interval_sec": cfg.interval_sec,
                    "params": cfg.params,
                    "warn_above": cfg.warn_above,
                    "critical_above": cfg.critical_above,
                }
            )
        return items

File: internal/collector/runner.go
package collector

import (
	"sync"
	"time"

	"github.com/pagbest154-cmd/system-monitor/internal/config"
	"github.com/pagbest154-cmd/system-monitor/internal/diskdiscovery"
	"github.com/pagbest154-cmd/system-monitor/internal/storage"
	"github.com/shirou/gopsutil/v4/cpu"
)

type UpdateCallback func(map[string]interface{})

type Collector struct {
	store        *storage.MetricStore
	onUpdate     UpdateCallback
	configLoader func() (*config.SensorsFile, error)
	stop         chan struct{}
	done         chan struct{}
	mu           sync.Mutex
	cfg          *config.SensorsFile
	sensors      map[string]Sensor
	lastPoll     map[string]float64
	latest       map[string]map[string]interface{}
}

func NewCollector(store *storage.MetricStore, onUpdate UpdateCallback, cfg *config.SensorsFile, loader func() (*config.SensorsFile, error)) *Collector {
	if loader == nil {
		loader = func() (*config.SensorsFile, error) { return config.LoadSensorsConfig("") }
	}
	c := &Collector{
		store: store, onUpdate: onUpdate, configLoader: loader,
		stop: make(chan struct{}), done: make(chan struct{}),
		sensors: map[string]Sensor{}, lastPoll: map[string]float64{},
		latest: map[string]map[string]interface{}{},
	}
	if cfg != nil {
		c.cfg = cfg
	} else {
		c.cfg, _ = loader()
	}
	c.ReloadConfig(nil)
	return c
}

func (c *Collector) ReloadConfig(cfg *config.SensorsFile) {
	if cfg != nil {
		c.cfg = cfg
	} else {
		c.cfg, _ = c.configLoader()
	}
	sensors := map[string]Sensor{}
	configuredPaths := map[string]bool{}
	for _, item := range c.cfg.Sensors {
		if item.Type == "system.disk_usage" && item.Params != nil {
			if p, ok := item.Params["path"].(string); ok {
				configuredPaths[p] = true
			}
		}
	}
	for _, item := range c.cfg.Sensors {
		if !item.Enabled || !item.IsSupportedOnPlatform() {
			continue
		}
		sensor, err := CreateSensor(item)
		if err != nil {
			continue
		}
		sensors[item.ID] = sensor
	}
	if c.cfg.Settings.AutoDiscoverDisks {
		for _, item := range diskdiscovery.DiscoverDiskSensors() {
			path, _ := item.Params["path"].(string)
			if configuredPaths[path] {
				continue
			}
			sensor, err := CreateSensor(item)
			if err != nil {
				continue
			}
			sensors[item.ID] = sensor
		}
	}
	c.mu.Lock()
	c.sensors = sensors
	c.mu.Unlock()
}

func (c *Collector) Start() {
	go c.run()
}

func (c *Collector) Stop() {
	close(c.stop)
	<-c.done
}

func (c *Collector) run() {
	_, _ = cpu.Percent(0, false)
	lastCleanup := 0.0
	ticker := time.NewTicker(500 * time.Millisecond)
	defer ticker.Stop()
	defer close(c.done)
	for {
		select {
		case <-c.stop:
			return
		case <-ticker.C:
			now := float64(time.Now().UnixNano()) / 1e9
			updates := map[string]interface{}{}
			c.mu.Lock()
			sensors := make(map[string]Sensor, len(c.sensors))
			for k, v := range c.sensors {
				sensors[k] = v
			}
			c.mu.Unlock()
			for sensorID, sensor := range sensors {
				cfg := sensor.Config()
				interval := c.intervalFor(cfg)
				last := c.lastPoll[sensorID]
				if now-last < float64(interval) {
					continue
				}
				reading := c.safeRead(sensor)
				if c.store != nil {
					_ = c.store.Insert(sensorID, reading.Value, reading.Status, now)
				}
				payload := reading.ToMap()
				payload["name"] = cfg.Name
				payload["unit"] = cfg.Unit
				payload["type"] = cfg.Type
				payload["ts"] = now
				c.mu.Lock()
				c.latest[sensorID] = payload
				c.mu.Unlock()
				updates[sensorID] = payload
				c.lastPoll[sensorID] = now
			}
			if len(updates) > 0 && c.onUpdate != nil {
				c.onUpdate(updates)
			}
			if c.store != nil && now-lastCleanup > 3600 {
				_, _ = c.store.Cleanup(c.cfg.Settings.RetentionDays)
				lastCleanup = now
			}
		}
	}
}

func (c *Collector) intervalFor(cfg config.SensorConfig) int {
	if cfg.IntervalSec != nil {
		return *cfg.IntervalSec
	}
	return c.cfg.Settings.DefaultIntervalSec
}

func (c *Collector) safeRead(sensor Sensor) SensorReading {
	reading := sensor.Read()
	if reading.Status == "" {
		reading.Status = "ok"
	}
	return reading
}

func (c *Collector) GetLatestSnapshot() map[string]map[string]interface{} {
	c.mu.Lock()
	defer c.mu.Unlock()
	out := make(map[string]map[string]interface{}, len(c.latest))
	for k, v := range c.latest {
		out[k] = v
	}
	return out
}

func (c *Collector) GetSensorConfigs() []config.SensorConfig {
	configs := append([]config.SensorConfig{}, c.cfg.Sensors...)
	configuredPaths := map[string]bool{}
	for _, cfg := range configs {
		if cfg.Type == "system.disk_usage" && cfg.Params != nil {
			if p, ok := cfg.Params["path"].(string); ok {
				configuredPaths[p] = true
			}
		}
	}
	for _, item := range diskdiscovery.DiscoverDiskSensors() {
		path, _ := item.Params["path"].(string)
		if configuredPaths[path] {
			continue
		}
		configs = append(configs, item)
	}
	return configs
}

func (c *Collector) GetSettings() map[string]interface{} {
	return map[string]interface{}{
		"retention_days":       c.cfg.Settings.RetentionDays,
		"default_interval_sec": c.cfg.Settings.DefaultIntervalSec,
		"auto_discover_disks":  c.cfg.Settings.AutoDiscoverDisks,
	}
}

func (c *Collector) GetActiveSensorIDs() map[string]bool {
	c.mu.Lock()
	defer c.mu.Unlock()
	out := make(map[string]bool, len(c.sensors))
	for id := range c.sensors {
		out[id] = true
	}
	return out
}

func (c *Collector) BuildSensorMeta() []map[string]interface{} {
	active := c.GetActiveSensorIDs()
	items := make([]map[string]interface{}, 0)
	for _, cfg := range c.GetSensorConfigs() {
		if !active[cfg.ID] {
			continue
		}
		item := map[string]interface{}{
			"id": cfg.ID, "name": cfg.Name, "type": cfg.Type, "unit": cfg.Unit,
			"enabled": cfg.Enabled, "params": cfg.Params,
		}
		if cfg.IntervalSec != nil {
			item["interval_sec"] = *cfg.IntervalSec
		}
		if cfg.WarnAbove != nil {
			item["warn_above"] = *cfg.WarnAbove
		}
		if cfg.CriticalAbove != nil {
			item["critical_above"] = *cfg.CriticalAbove
		}
		items = append(items, item)
	}
	return items
}
